### ISSUE TITLE
Differences with/without packrat (-s2 simplification issue)

### ISSUE THREAD

**Issue by ChrisHixon** [ChrisHixon/chpeg#25]
Testing the attached kotlin grammar with the latest chpeg gives a slightly different output with and without -packrat:
```
>/usr/bin/time ../../chpeg-chris0/util/chpeg kotlin-perf-packrat.chpeg kotlin.kt -packrat
 OFFSET   LEN     ID NC FLG IDENT "DATA"
     0  49029      0 540 R-- file "/*\n * Copyright 2000-2018 JetBrains ..."
   210     39      5  2 ---   packageHeader "package org.jetbrains.kotlin.renderer..."
...
  1427      1     26  0 ---       MULT "*"
  1430    201     10  1 ---   topLevelObject "internal class DescriptorRendererImpl..."
  1430    148     39  6 ---     classDeclaration "internal class DescriptorRendererImpl..."
  1430      8    312  0 ---       INTERNAL "internal"
  1439      5     43  0 ---       CLASS "class"
  1445     22    342  0 --L       Identifier "DescriptorRendererImpl"
  1467     50     52  3 ---       classParameters "(\n    val options: DescriptorRendere..."
...
 49029      0      9  0 ---   EOF ""
0.11user 0.05system 0:00.17elapsed 97%CPU (0avgtext+0avgdata 76164maxresident)k
0inputs+0outputs (0major+50252minor)pagefaults 0swaps
```
```
>/usr/bin/time ../../chpeg-chris0/util/chpeg kotlin-perf-packrat.chpeg kotlin.kt
 OFFSET   LEN     ID NC FLG IDENT "DATA"
     0  49029      0 540 R-- file "/*\n * Copyright 2000-2018 JetBrains ..."
   210     39      5  2 ---   packageHeader "package org.jetbrains.kotlin.renderer..."
   210      7     21  0 ---     PACKAGE "package"
...
  1427      1     26  0 ---       MULT "*"
  1430    148     39  6 ---   classDeclaration "internal class DescriptorRendererImpl..."
  1430      8    312  0 ---     INTERNAL "internal"
  1439      5     43  0 ---     CLASS "class"
  1445     22    342  0 --L     Identifier "DescriptorRendererImpl"
  1467     50     52  3 ---     classParameters "(\n    val options: DescriptorRendere..."
...
49029      0      9  0 ---   EOF ""
14.38user 0.00system 0:14.39elapsed 99%CPU (0avgtext+0avgdata 16408maxresident)k
0inputs+0outputs (0major+3828minor)pagefaults 0swaps
```
[kotlin-perf-packrat.chpeg.zip](https://github.com/ChrisHixon/chpeg/files/8964535/kotlin-perf-packrat.chpeg.zip)

_Originally posted by @mingodad in https://github.com/ChrisHixon/chpeg/issues/23#issuecomment-1163998765_

**Comment by ChrisHixon** [comment]
This seems to be an issue with -s2 (new default AST simplification) and packrat. -s1 packrat/no-packrat ASTs are identical. It appears -s2 packrat version contains things that should be simplified. I'll look into it. Thanks for the report.

**Comment by ChrisHixon** [comment]
This should be fixed with 4d43d84eefe618d754fae9ae03c4bc8a56758b80

@mingodad, can you confirm this fixes the issue?

**Comment by ChrisHixon** [comment]
No problem, thanks for the report. I'll close this issue now. If you find anything else, please post another issue.

This one took a while to figure out! Removing excess trace info used to help figure this out: 1059c25af7d8ed306634e4fe872c671093b57f01
